### PR TITLE
weather-splitter: allow for more flexible output files

### PR DESCRIPTION
**Pull request by uhager**
Use python-style formatting for more flexible output files when splitting files in a directory hierarchy.

### CODE REVIEW

File: weather_sp/splitter_pipeline/file_name_utils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        input_base_dir: used if out_pattern does not contain any '{}' substitutions.
+            The output file is then created by replacing this part of the input name
+            with the output pattern.
+    '''
+    file_ending = ''
+    for ending in _WEATHER_FILE_ENDINGS:
+        if filename.endswith(ending):
+            file_ending = ending
+            filename = filename[:-len(ending)]
+            break
```
Comment (alxmrs) [inline weather_sp/splitter_pipeline/file_name_utils.py:41]: Consider using https://docs.python.org/3/library/os.path.html#os.path.splitext

File: weather_sp/splitter_pipeline/file_name_utils.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import logging
+import os
+import typing as t
+
+_WEATHER_FILE_ENDINGS = ('.nc', '.cd', '.grib', '.grb', '.grb2', '.grib2')
```
Comment (alxmrs) [inline weather_sp/splitter_pipeline/file_name_utils.py:5]: What about gzipped grib files? e.g. we see files with the extension `bz2` (and no other extension). Is there a way to process this without including a finite list of extensions?

**Review by alxmrs — CHANGES_REQUESTED** [review]
Thanks for your patch! I left some feedback. I think the design needs to be iterated on a bit more.

File: weather_sp/splitter_pipeline/pipeline.py [hunk 3]
```diff
@@ -66,12 +67,15 @@ def run(argv: t.List[str], save_main_session: bool = True):
...
+    parser.add_argument('-o', '--output-template', type=str, required=True,
+                        help='Template specifying path to output files. '
+                             'Either a directory that will replace the common path of the '
+                             'input_pattern, or a template using python-style formatting substitution'
+                             'of input directory names.'
+                             'For `input_pattern a/b/c/**` and file `a/b/c/file.nc`,'
+                             '`output_template /x/y/z` will create'
+                             'output files like `/x/y/z/c/file_shortname.nc`, while a template'
+                             'with formating `/somewhere/{1}-{0}.` will give `somewhere/c-file.shortname.nc`'
                         )
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/splitter_pipeline/pipeline.py]: Updated to reflect separate flags for directory vs template

File: weather_sp/README.md [hunk 4]
```diff
@@ -19,9 +19,7 @@ Split weather data file into files by variable.
 _Common options_:
 
 * `-i, --input-pattern`: Pattern for input weather data.
-* `-o, --output-dir`: Path to base folder for output files. This directory will replace the common path of the
-  input-pattern. For `input-pattern a/b/c/**` and
-  `output-dir /x/y/z` a file `a/b/c/file.nc` will createoutput files like `/x/y/z/c/file.nc_shortname.nc`
+* `-o, --output-template`: Path to base folder for output files, or template using Python formating, see [Output section](#output) below.
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/README.md]: Done. I now have two mutually exclusive flags.

File: weather_sp/README.md [hunk 5]
```diff
@@ -75,22 +73,45 @@ On the other hand, to specify a specific pattern use
...
---output-dir 'gs://test-output/splits'
+--output-template 'gs://test-output/splits/{2}.{1}.{0}T00.'
 ```
 
-For a file `gs://test-input/era5/2020/01/01.nc` the output file pattern is
-`gs://test-output/splits/2020/01/01.nc` and if the temperature is a variable in that data, the output file for that
-split will be `gs://test-output/splits/2020/01/01.nc_t.nc`
+For a file `gs://test-input/era5/2020/02/01.nc` the output file pattern is
+`gs://test-output/splits/2020.02.01T00.` and if the temperature is a variable in that data, the output file for that
+split will be `gs://test-output/splits/2020/01/01.t.nc`
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/README.md]: It went somewhere into the depth of my forgetfulness... added it.

File: weather_sp/README.md [hunk 6]
```diff
@@ -75,22 +73,45 @@ On the other hand, to specify a specific pattern use
...
+The substitutions are based on the directory structure of the input file, where each `{<x>}` stands for
+one directory name, counting backwards from the end, i.e. the file name is `{0}`, the immediate
+directory in which it is located is `{1}`.
 
 Example:
 
 ```bash
 --input-pattern 'gs://test-input/era5/2020/**' \
---output-dir 'gs://test-output/splits'
+--output-template 'gs://test-output/splits/{2}.{1}.{0}T00.'
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/README.md]: I explained that above, the numbers count backwards from the filename. If there are a variable number of directories, then it depends what those directories are. I don't think there is another way of doing this that would avoid this issue if the directory structure is different for different files. I considered flipping the order and counting from the root level directory up with the filename having the highest number but decided against: for deeper directory hierarchies, usually the last couple of directories are what varies and is interesting in this case. It's also easier to count, if the directories of interest are a few levels down a 10+ level path it's easier to use {0}, {1}, {2} than to use {12},{13},{14} and still get it right. Also, if we flip and now the number of directories varies, there is a larger chance that we get numbers that are higher than the number of directories for a specific file. Alternatively,and maybe worse because harder to debug than a simple crash, we might lose the important parts, like the actual file name.

In summary, I don't think there is a way around this issue, and if the directory structure doesn't fit in a single template, it's probably not a good idea to use a single template, and rather use output-dir to keep the information that is encoded in those directories.

File: weather_sp/README.md [hunk 7]
```diff
@@ -75,22 +73,45 @@ On the other hand, to specify a specific pattern use
...
+
+For a file `gs://test-input/era5/2020/02/01.nc` the output file pattern is
+`gs://test-output/splits/2020/02/01_` and if the temperature is a variable in that data, the output file for that
+split will be `gs://test-output/splits/2020/02/01_t.nc`
+
+### Output template with Python-style formating
+Using Python-style substitution (e.g. `{1}`) allows for more flexibility when creating the output files.
+The substitutions are based on the directory structure of the input file, where each `{<x>}` stands for
+one directory name, counting backwards from the end, i.e. the file name is `{0}`, the immediate
+directory in which it is located is `{1}`.
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/README.md]: I'm not sure how else to phrase this. I flipped the numbers in the example to illustrate that in '{x}' the 'x' represents the level starting with 0 for the filename.

File: weather_sp/splitter_pipeline/file_name_utils.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import logging
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py:15]: Done.

File: weather_sp/splitter_pipeline/file_name_utils.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+class OutFileInfo(t.NamedTuple):
+    file_name_base: str
+    ending: str
+
+    def __str__(self):
+        return f'{self.file_name_base}/*/{self.ending}'
+
+
+def get_output_file_base_name(filename: str, out_pattern: str, input_base_dir: str) -> OutFileInfo:
+    '''
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py]: Changes to """. I don't understand what you mean by 'no return'.

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 1 above)
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py]: Done.

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 2 above)
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py:19]: Splitter doesn't currently support zipped or tarred files, I think that's a separate feature request and would require more thought. I think to support that we'd likely unzip first anyway, and I don't know if we should really support rezipping the split files, would we then have to support every posible zip format?

File: weather_sp/splitter_pipeline/file_name_utils.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    Construct the base output file name by applying the out_pattern to the
+    filename.
+
+    Example:
+        filename = 'gs://my_bucket/data_to_split/2020/01/21.nc'
+        out_pattern = 'gs://my_bucket/splits/{2}-{1}-{0}_old_data.'
+        resulting output base = 'gs://my_bucket/splits/2020-01-21_old_data.'
+        resulting file ending = '.nc'
```
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py:44]: I think this is overkill here: we already have a separate test, and I think the example is actually more readable in text than it would be in code.

File: weather_sp/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_sp/README.md]: I'm happy to document that users milage may vary based on their input pattern. One idea that comes to mind: Have the number of items in the output pattern depend on the format of the input pattern. For now, we can keep this approach in mind for another time.

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_sp/splitter_pipeline/file_name_utils.py]: The summary line and the `"""` should be on the same line: There should be no `\n` between the block quote statement and the first sentence. 

Style guide for reference (see the example in the section below): https://google.github.io/styleguide/pyguide.html#381-docstrings

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_sp/splitter_pipeline/file_name_utils.py:44]: I agree that executing these tests are overkill. However, I do like the convention that the examples are python code, and I am ok with your preference.

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 2 above)
Comment (alxmrs) [reply weather_sp/splitter_pipeline/file_name_utils.py:19]: As far as opening files, I just did an experiment and it looks like Apache Beam filesystems does a good job of decompressing blobs in a wide range of formats. We just need to support the URI parsing so we can accept the files. 
A general principle that the weather tools have attempted to follow so far is to make an educated guess for processing, but ultimately give users control over processing. If the splitter had some way to override the allowed extensions, then it could be used in more cases.

File: weather_sp/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/README.md]: I think making this even more flexible requires more design. I also worry that we end up making this too error-prone that we accept all kinds of input and output pattern and create output files that may be unexpected. I think in some cases (like changing input directory depth) it may be better to fail loudly and then offer a flag for the user to override and run anyway. But I'd rather do that in a separate CL.

File: weather_sp/splitter_pipeline/file_name_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (uhager) [reply weather_sp/splitter_pipeline/file_name_utils.py]: Done.

**Review by alxmrs — APPROVED** [review]
LGTM
